docs/framework/wcf/samples/custom-lifetime.md

```diff
@@ -1,219 +1,227 @@
-}  
-  
-class CustomLeaseExtension : IExtension<InstanceContext>, ICustomLease  
-{  
-}  
-```  
-  
- When WCF invokes the <xref:System.ServiceModel.Dispatcher.IInstanceContextProvider.IsIdle%2A> method in the <xref:System.ServiceModel.Dispatcher.IInstanceContextProvider> implementation this call is routed to the <xref:System.ServiceModel.Dispatcher.IInstanceContextProvider.IsIdle%2A> method of the `CustomLeaseExtension`. Then the `CustomLeaseExtension` checks its private state to see whether the <xref:System.ServiceModel.InstanceContext> is idle. If it is idle it returns `true`. Otherwise, it starts a timer for a specified amount of extended lifetime.  
-  
-```  
-public bool IsIdle  
-{  
-  get  
-  {  
-    lock (thisLock)  
-    {  
-      if (isIdle)  
-      {  
-        return true;  
-      }  
-      else  
-      {  
-        StartTimer();  
-        return false;  
-      }  
-    }  
-  }  
-}  
-```  
-  
- In the timer’s `Elapsed` event the callback function in the Dispatcher is called in order to start another clean up cycle.  
-  
-```  
-void idleTimer_Elapsed(object sender, ElapsedEventArgs args)  
-{  
-    idleTimer.Stop();  
-    isIdle = true;    
-    callback(owner);  
-}  
-```  
-  
- There is no way to renew the running timer when a new message arrives for the instance being moved to the idle state.  
-  
- The sample implements <xref:System.ServiceModel.Dispatcher.IInstanceContextProvider> to intercept the calls to the <xref:System.ServiceModel.Dispatcher.IInstanceContextProvider.IsIdle%2A> method and route them to the `CustomLeaseExtension`. The <xref:System.ServiceModel.Dispatcher.IInstanceContextProvider> implementation is contained in `CustomLifetimeLease` class. The <xref:System.ServiceModel.Dispatcher.IInstanceContextProvider.IsIdle%2A> method is invoked when WCF is about to release the service instance. However, there is only one instance of a particular `ISharedSessionInstance` implementation in the ServiceBehavior’s <xref:System.ServiceModel.Dispatcher.IInstanceContextProvider> collection. This means there is no way of knowing the <xref:System.ServiceModel.InstanceContext> being closed at the time WCF checks the <xref:System.ServiceModel.Dispatcher.IInstanceContextProvider.IsIdle%2A> method. Therefore this sample uses thread locking to serialize requests to the <xref:System.ServiceModel.Dispatcher.IInstanceContextProvider.IsIdle%2A> method.  
-  
+> The setup procedure and build instructions for this sample are located at the end of this article.
+
+## Shared instancing
+
+WCF offers several instancing modes for your service instances. The shared instancing mode covered in this article provides a way to share a service instance between multiple channels. Clients can contact a factory method in the service and create a new channel to start communication. The following code snippet shows how a client application creates a new channel to an existing service instance:
+
+```csharp
```

Comment:
This section needs cleaning up as the description doesn't match the code snippet. I'll update this either later today or more likely tomorrow.

Reply:
Sounds good @mconnew.

Reply:
@mconnew following up on the updated description. Can you please help? I'd like to merge this PR soon.

Reply:
The code should be this:

```c#
// Create a header for the shared instance id
MessageHeader shareableInstanceContextHeader = MessageHeader.CreateHeader(
        CustomHeader.HeaderName,
        CustomHeader.HeaderNamespace,
        Guid.NewGuid().ToString());

// Create the channel factory
ChannelFactory<IEchoService> channelFactory =
    new ChannelFactory<IEchoService>("echoservice");

// Create the first channel
IEchoService proxy = channelFactory.CreateChannel();

// Call an operation to create shared service instance
using (new OperationContextScope((IClientChannel)proxy))
{
    OperationContext.Current.OutgoingMessageHeaders.Add(shareableInstanceContextHeader);
    Console.WriteLine("Service returned: " + proxy.Echo("Apple"));
}

((IChannel)proxy).Close();

// Create the second channel
IEchoService proxy2 = channelFactory.CreateChannel();

// Call an operation using the same header which will reuse the shared service instance
using (new OperationContextScope((IClientChannel)proxy2))
{
    OperationContext.Current.OutgoingMessageHeaders.Add(shareableInstanceContextHeader);
    Console.WriteLine("Service returned: " + proxy2.Echo("Apple"));
}
```